### PR TITLE
Notification Rework

### PR DESCRIPTION
**Pull request by mharasic**
WIP

### CODE REVIEW

File: backend/api/helper/uploader.py [hunk 1]
```diff
@@ -37,15 +37,9 @@ def save_manifest_to_db(manifest):
                     entry['archived'] = False
                 entry['authors'] = sorted(list(set(entry['authors'])))
                 entry['tags'] = sorted(entry['tags'])
-                g.projects.insert(entry)
+                projects.append(entry)
```
Comment (henderkes) [inline backend/api/helper/uploader.py:40]: erm, the uploaded projects will no longer be added to the database?

File: backend/__init__.py [hunk 2]
```diff
@@ -116,21 +123,6 @@ class Role(DB.Document, RoleMixin):
     description = DB.StringField(max_length=255)
 
 
-class Notification(DB.EmbeddedDocument):
```
Comment (henderkes) [inline backend/__init__.py:119]: ?

File: backend/api/notifications.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import uuid
+
+
+notifications = Blueprint('api_notifications', __name__)
+
+
+def add_notification(creator, userlist, project_id, operation, reason='', saved_search_id=''):
+    date = time.strftime("%Y-%m-%d %H:%M:%S")
+    for user in userlist:
+        if user not in (creator):
```
Comment (henderkes) [inline backend/api/notifications.py]: this is a poor equality check

File: backend/api/notifications.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            g.notifications.insert_one({
+                '_id': str(uuid.uuid4()),
+                'creator': creator,
+                'user_id': user,
```
Comment (henderkes) [inline backend/api/notifications.py:19]: do you think this is a good idea? notifications are pretty much made to be stored as EmbeddedDocuments.

**Review by henderkes — COMMENTED** [review]
I don't think this is a good idea all around to be honest. Splitting the files into different blueprints is great, but I'm not a fan of the rest

File: backend/api/helper/uploader.py (hunk 1 above)
Comment (mharasic) [reply backend/api/helper/uploader.py:40]: They are inserted in  projects.py

File: backend/api/notifications.py (hunk 4 above)
Comment (mharasic) [reply backend/api/notifications.py:19]: Feel free to make them as Embedded documents... Using pymongo only for convenience

File: backend/__init__.py (hunk 2 above)
Comment (mharasic) [reply backend/__init__.py:119]: USing pymongo !?!

File: backend/api/notifications.py (hunk 4 above)
Comment (henderkes) [reply backend/api/notifications.py:19]: They already were, that's why I'm confused you put them in their own collection now, this just seems confusing to me

File: backend/api/projects.py [hunk 5]
```diff
@@ -46,27 +35,40 @@ def add_projects():
...
             else:
                 raise ApiException("Wrong content header" +
                                    "and no files attached", 400)
-
-            g.rerun_saved_searches()
-            return jsonify(return_ids)
+            project = projects[0]
+            if 'comments' not in project:
+                project['comments'] = []
+            g.projects.insert(project)
```
Comment (henderkes) [inline backend/api/projects.py]: only the first is saved, this seems like a huge mess

File: backend/__init__.py [hunk 6]
```diff
@@ -273,9 +274,10 @@ def handle_invalid_usage(error):
 def handle_insufficient_permission(error):
     """ Handler for insufficient permission to access a method.
         This is not the error handler for insufficient permission to update
-        a project or user.
+        a project or user. It delivers a 404 as users should not see endpoints,
+        where they don't have permissions to.
```
Comment (henderkes) [inline backend/__init__.py:278]: this handler is for cases where the user knows the endpoint but doesn't have permission to use it.
the other errorhandler is for cases where the user should not be aware of the endpoint

Comment (mharasic) [reply backend/__init__.py:278]: Thats the 404 I'm not admin handler.... Strange

File: backend/api/projects.py (hunk 5 above)
Comment (mharasic) [reply backend/api/projects.py]: Fixed..... Was trappend in a different understanding of code.... In that actual case we have unreachable code which was planned to work dirrently...

**Pull request merged**